[Gareth Koh] iP

garethkoh opened the pull request:
## Duke
> "your mind is for having ideas"

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use
All you need to do is,

1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 💯 

its **FREE**

tasklist:
- [x] done
- [ ] not done

here is the main method:

```java
public class Duke {

    public static void main(String[] args) {
        // load in the TaskList
        TaskList taskList = new TaskList(new Storage());
        Parser parser = new Parser(taskList);
        Ui.initialize(parser);
        Storage.save(taskList);

    }
```

Commit:
Add Gradle support

garethkoh committed:
Level 1: Greet, Echo, Exit.

garethkoh committed:
ignore .class files

garethkoh committed:
Level 2: Add, List.

garethkoh committed:
Level 3: Mark as done.

garethkoh committed:
moved Task class to a separate file

garethkoh committed:
added subclasses of Task

garethkoh committed:
Level 4: ToDos, Events, Deadlines.

garethkoh committed:
Level 6: Delete

garethkoh committed:
Automated text UI testing

garethkoh committed:
Level 5. Handle errors. all errors handled.

garethkoh committed:
Reformat code to coding standard

garethkoh committed:
Update expected test result

garethkoh committed:
Fix coding standard violations

garethkoh committed:
Add Save functionality

garethkoh committed:
Ignore data.txt file that contains users list

garethkoh committed:
Add date time formatting

garethkoh committed:
Remove unnecessary imports

garethkoh committed:
Resolve merge conflicts

garethkoh committed:
Merge branch 'branch-Level-8'

Korochii commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+        Scanner sc = new Scanner(System.in);
+        ArrayList<Task> toDoList = new ArrayList<>();
+        try {
+
```
Might want to delete this line (line 32) since it feels abit unneccesary

Korochii commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+                    Task t = toDoList.get(i);
+                    System.out.printf("%d. " + t.toString(), i + 1);
+                }
+            } else if (split[0].equals("help")) {
+                System.out.println("try the following commands:\n"
+                        + "   list (list out all tasks in your todo list)\n"
+                        + "   todo <task> (add a basic task to your todo list)\n"
+                        + "   deadline <task> /by <date> (add a task with a deadline to your todo list)\n"
+                        + "   event <task> /at <date> (add a new event with its corresponding date to your todo list)\n"
+                        + "   mark <task index> (mark that specific task as completed)\n"
+                        + "   unmark <task index> (remove mark from specific task)\n"
+                        + "   delete <task index> (remove task from todo list)\n"
+                        + "   bye (close application)");
```
Interesting addition to Duke that was not specified in the basic requirements (and I did not think of implementing this too!)
Good job 😃

Korochii commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+                } catch (NumberFormatException e) {
+                    System.out.println("unmark description must be an integer!");
+                }
+
+
```
Might want to delete these empty lines too so that the formatting is abit more consistent

Korochii commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+                } catch (NumberFormatException e) {
+                    System.out.println("mark description must be an integer!");
+                }
+
```
Might want to delete this empty line too so that the formatting is abit more consistent

Korochii commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+                        // Take the substring of user input after deadline
+                        String item = userInput.substring(9);
+                        // Divide the substring into task name and deadline
+                        String[] divide = item.split("/");
```
In this case, divide should be in plural form since it is an array of strings

Korochii commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+                        // Take the substring of user input after event
+                        String item = userInput.substring(6);
+                        // Divide the substring into task name and date
+                        String[] divide = item.split("/");
```
In this case, divide should be in plural form since it is an array of strings

Korochii reviewed (commented):
Overall, most of the coding standards are being adhered to (which is why I had to nitpick a few minor issues :sweat_smile:)

Good job and keep up the good work!

cheehongw commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public String toString() {
+        if (super.getStatus() == 1) {
+            return "[E][X] " + super.getName() + " (at: "
+                    + time.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.FULL)) + ")\n";
+        } else {
+            return "[E][ ] " + super.getName() + " (at: "
+                    + time.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.FULL)) + ")\n";
+        }
```
perhaps instead of repeating the return string twice to handle the different status of the Task, you can first do something like:

```java
public String toString() {
        String statusString = super.getStatus() == 1 ? "[E][X]" : "[E][ ]";
        return statusString + super.getName() + " (at: "
                    + time.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.FULL)) + ")\n";
        

```

what do you think 🤔 ?

cheehongw commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,38 @@
+class Task implements Serializable {
+
+    // 1 indicates done and 0 indicates not done
+    private int status;
```
why not use a boolean?

cheehongw commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+        }
+        System.out.println("goodbye!");
+        try {
+            FileOutputStream writer = new FileOutputStream("../../../data/data.txt");
```
Try to avoid "magic values" such as this `"../../../data/data.txt"` filepath here.

Perhaps declare it as a constant instead?

cheehongw commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,223 @@
+                }
+
+            } else if (split[0].equals("delete")) {
+                try {
+                    // Checks if description is empty
+                    int toDelete = Integer.parseInt(split[1]);
+                    try {
+                        Task t = toDoList.get(toDelete - 1);
+                        toDoList.remove(toDelete - 1);
+                        System.out.printf("task removed:\n%s", t.toString());
+                        System.out.printf("you now have %d tasks\n", toDoList.size());
+                    } catch (IndexOutOfBoundsException e) {
+                        System.out.println(("the index you have entered does not exist!"));
+                    }
+                } catch (IndexOutOfBoundsException e) {
+                    System.out.println("delete description cannot be empty!");
+                } catch (NumberFormatException e) {
+                    System.out.println("delete description must be an integer!");
+                }
```
try not to nest try-catch within try-catch

cheehongw reviewed (commented):
The `main` method in `Duke.java` is too long and difficult to follow. It is deeply nested with try-catch and if-else blocks.

Maybe try and abstract out the error handling parts to other methods to reduce the number of try-catch blocks? 🤔

garethkoh committed:
Refactor code to OOP standard

garethkoh committed:
Fix printed text formatting

garethkoh committed:
Fix unchecked warning in storage class

garethkoh committed:
Refactor code into package

garethkoh committed:
Add colour to output text

garethkoh committed:
Merge branch 'add-gradle-support'

garethkoh committed:
Refactor code

garethkoh committed:
Add Gradle support

garethkoh committed:
Add JUnit tests

garethkoh committed:
Refactor code for Storage class

garethkoh committed:
Refactor code in Storage class

garethkoh committed:
Ignore .jar files

garethkoh committed:
Add javadoc comments

garethkoh committed:
Change method to fit coding standard

garethkoh committed:
Improve code quality of Parser class

garethkoh committed:
Add javadoc comments for all Command classes

garethkoh committed:
Add find command

garethkoh committed:
Bug fix for program crashing when date format is wrong

garethkoh committed:
Merge branch 'branch-A-JavaDoc'

garethkoh committed:
Merge branch 'branch-Level-9'

garethkoh committed:
Add JavaDoc for CommandFind class

garethkoh committed:
Format text output

garethkoh committed:
Update build.gradle file to include javaFX

garethkoh committed:
JavaFX Tutorial

garethkoh committed:
Implement GUI

garethkoh committed:
Remove main method in Duke class

garethkoh committed:
Close javafx application on bye command

bye command only prints "goodbye!" and does not shut down Duke as
expected.

it is unintuitive if the bye command has been sent but the application
is still running.

Let's make Duke shuts down 3 seconds after bye command so that the user
can see any messages sent by duke before shutting down.

garethkoh committed:
Refactor code to more OOP

garethkoh committed:
Add assertions

garethkoh committed:
Add comments

garethkoh committed:
Refactor code to improve code quality

Code contained nested try-catch statements and booleans that do not
sound like booleans.

Code was not readable.

Let's replace some try-catch statements with if statements and extract
individual try-catch blocks as methods where not possible to remove
them. Also, rename booleans to sound like booleans.

This way the code is more readable.

garethkoh committed:
Add Junit test for deadline command

garethkoh committed:
Reduce delay for closing application upon bye command

garethkoh committed:
Merge pull request #2 from garethkoh/branch-A-Assertions

Branch a assertions

garethkoh committed:
Merge branch 'master' into branch-A-CodeQuality

garethkoh committed:
Merge pull request #3 from garethkoh/branch-A-CodeQuality

Branch a code quality

garethkoh committed:
Add extension C-DetectDuplicates

garethkoh committed:
Credit external sources

garethkoh committed:
Change initialisation message

garethkoh committed:
Update build.gradle

jar file not working cross platform

garethkoh committed:
Add functionality to save upon any modification to list

garethkoh committed:
Create Ui.png

garethkoh committed:
Update Ui.png

garethkoh committed:
Update Ui.png

garethkoh committed:
Add Ui.png

garethkoh committed:
Update Ui.png

garethkoh committed:
Update README.md

garethkoh committed:
Edit CommandHelp

garethkoh committed:
Update build.gradle
